dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1324,6 +1330,53 @@ bool RtpsUdpDataLink::requires_inline_qos(const GUIDSeq_var& peers)
 
 bool RtpsUdpDataLink::force_inline_qos_ = false;
 
+void
+RtpsUdpDataLink::RtpsWriter::send_heartbeats(const DCPS::MonotonicTimePoint& now)
+{
+  OPENDDS_VECTOR(TransportQueueElement*) pendingCallbacks;
+
+  RtpsUdpDataLink_rch link = link_.lock();
+  if (!link) {
+    return;
+  }
+
+  RtpsUdpInst& cfg = link->config();
+
+  {
+    ACE_GUARD(ACE_Thread_Mutex, g2, heartbeat_mutex_);
+    if (expected_acks_) {
+      // One more readers did not ack since the last heartbeat.
+      heartbeat_period_ *= cfg.heartbeat_backoff_factor_;
+    } else if (last_ack_ != MonotonicTimePoint::zero_value && last_heartbeat_ != MonotonicTimePoint::zero_value) {
+      heartbeat_period_ = (last_ack_ - last_heartbeat_) * cfg.heartbeat_safety_factor_;
+    }
+    heartbeat_period_ = std::min(heartbeat_period_, std::max(cfg.heartbeat_period_maximum_, cfg.heartbeat_period_minimum_));
+    heartbeat_period_ = std::max(heartbeat_period_, std::min(cfg.heartbeat_period_maximum_, cfg.heartbeat_period_minimum_));
+    expected_acks_ = 0;
+  }
+
+  MetaSubmessageVec meta_submessages;
+  send_and_gather_nack_replies(meta_submessages);
+  const bool keep_going = gather_heartbeats(pendingCallbacks, meta_submessages);
+  link->queue_or_send_submessages(meta_submessages);
+
+  for (size_t i = 0; i < pendingCallbacks.size(); ++i) {
+    pendingCallbacks[i]->data_dropped();
+  }
+
+  {
+    ACE_GUARD(ACE_Thread_Mutex, g2, heartbeat_mutex_);
+    if (keep_going) {
+      heartbeat_.cancel();
+      heartbeat_.schedule(heartbeat_period_);
```

Comment:
Hmm, we're canceling and rescheduling every heartbeat for every writer? What's the cancel / reschedule delay for SporadicTask? It's definitely non-zero for MultiTask, which is why I added the cancel_estimate_ field. Maybe you could use MultiTask with the constructed default of the max heartbeat period and just enable with the new heartbeat period every time? I'm not sure if it will be much more efficient, but it seems like it will be more robust to cancelation delays and eliminate the need for a cancel call here.

Reply:
I think the code it question no longer exists.  Don't know why github is not showing outdated.